src/tools/wasm-interp.cc

```diff
@@ -401,107 +401,103 @@ static interpreter::Result default_host_callback(const HostFunc* func,
-                                           PrintErrorCallback callback,
-                                           void* user_data) {
-  if (string_slice_eq_cstr(&import->field_name, "global")) {
-    switch (global->typed_value.type) {
-      case Type::I32:
-        global->typed_value.value.i32 = 666;
-        break;
+  wabt::Result ImportGlobal(Import* import,
+                            Global* global,
+                            const ErrorCallback& callback) override {
+    if (string_slice_eq_cstr(&import->field_name, "global")) {
+      switch (global->typed_value.type) {
+        case Type::I32:
+          global->typed_value.value.i32 = 666;
+          break;
 
-      case Type::F32: {
-        float value = 666.6f;
-        memcpy(&global->typed_value.value.f32_bits, &value, sizeof(value));
-        break;
-      }
+        case Type::F32: {
+          float value = 666.6f;
+          memcpy(&global->typed_value.value.f32_bits, &value, sizeof(value));
+          break;
+        }
 
-      case Type::I64:
-        global->typed_value.value.i64 = 666;
-        break;
+        case Type::I64:
+          global->typed_value.value.i64 = 666;
+          break;
 
-      case Type::F64: {
-        double value = 666.6;
-        memcpy(&global->typed_value.value.f64_bits, &value, sizeof(value));
-        break;
+        case Type::F64: {
+          double value = 666.6;
+          memcpy(&global->typed_value.value.f64_bits, &value, sizeof(value));
+          break;
+        }
+
+        default:
+          PrintError(callback, "bad type for host global import " PRIimport,
+                     PRINTF_IMPORT_ARG(*import));
+          return wabt::Result::Error;
       }
 
-      default:
-        print_error(callback, "bad type for host global import " PRIimport,
-                    PRINTF_IMPORT_ARG(*import));
-        return wabt::Result::Error;
+      return wabt::Result::Ok;
+    } else {
+      PrintError(callback, "unknown host global import " PRIimport,
+                 PRINTF_IMPORT_ARG(*import));
+      return wabt::Result::Error;
     }
+  }
 
-    return wabt::Result::Ok;
-  } else {
-    print_error(callback, "unknown host global import " PRIimport,
-                PRINTF_IMPORT_ARG(*import));
-    return wabt::Result::Error;
+ private:
+  void PrintError(const ErrorCallback& callback, const char* format, ...) {
+    WABT_SNPRINTF_ALLOCA(buffer, length, format);
+    callback(buffer);
   }
-}
+};
 
 static void init_environment(Environment* env) {
   HostModule* host_module =
       env->AppendHostModule(string_slice_from_cstr("spectest"));
-  host_module->import_delegate.import_func = spectest_import_func;
-  host_module->import_delegate.import_table = spectest_import_table;
-  host_module->import_delegate.import_memory = spectest_import_memory;
-  host_module->import_delegate.import_global = spectest_import_global;
+  host_module->import_delegate.reset(new SpectestHostImportDelegate());
```

Comment:
I've become used to see make_unique() used for this.  llvm implements it locally since its in C++14 only I think: http://llvm.org/docs/doxygen/html/STLExtras_8h_source.html